packages/openapi3/src/decorators.ts

```diff
@@ -32,3 +50,64 @@ export const $oneOf: OneOfDecorator = (
 export function getOneOf(program: Program, entity: Type): boolean {
   return program.stateMap(oneOfKey).get(entity);
 }
+
+const [getTagsMetadata, setTagsMetadata] = unsafe_useStateMap<
+  Type,
+  { [name: string]: OpenAPI3Tag }
+>(OpenAPI3Keys.tagsMetadata);
+
+export const $tagMetadata: TagMetadataDecorator = (
+  context: DecoratorContext,
+  entity: Namespace,
+  name: string,
+  tagMetadata?: TypeSpecValue,
+) => {
+  const tags = getTagsMetadata(context.program, entity);
+  if (tags && tags[name]) {
+    reportDiagnostic(context.program, {
+      code: "duplicate-tag",
+      format: { tagName: name },
+      target: context.getArgumentTarget(0)!,
+    });
+  }
+  let metadata: OpenAPI3Tag = { name };
+  if (tagMetadata) {
+    const [data, diagnostics] = typespecTypeToJson<OpenAPI3Tag & Record<ExtensionKey, unknown>>(
+      tagMetadata,
+      context.getArgumentTarget(0)!,
+    );
+    context.program.reportDiagnostics(diagnostics);
+    if (data === undefined) {
+      return;
+    }
+    validateAdditionalInfoModel(context, tagMetadata, data);
+    metadata = { ...data, name };
+  }
+
+  const newTags = { ...tags, [name]: metadata };
+  setTagsMetadata(context.program, entity, newTags);
+};
+
+export { getTagsMetadata };
+
+function validateAdditionalInfoModel(
+  context: DecoratorContext,
+  typespecType: TypeSpecValue,
+  data: OpenAPI3Tag & Record<`x-${string}`, unknown>,
+) {
+  const propertyModel = context.program.resolveTypeReference(
+    "TypeSpec.OpenAPI.TagMetadata",
+  )[0]! as Model;
+  const diagnostics: Diagnostic[] = [];
+  if (typeof typespecType === "object" && propertyModel) {
+    diagnostics.push(
```

Comment:
report? no need to add to an array

Reply:
it need report all error/warning once. so push array first, then call `context.program.reportDiagnostics(diagnostics);`